File: backend/routers/calendario.py
"""
Router de Calendário — visão de tempo cruzando todas as matérias.

Endpoints:
  GET    /calendario                  atividades respondidas + eventos numa janela
  POST   /calendario/eventos          marca um evento (prova, revisão, tarefa)
  DELETE /calendario/eventos/{id}     remove um evento

O passado vem de activity_results (o que o aluno realmente fez) e o futuro de
calendar_events (o que ele marcou). Provas são eventos com kind='prova' —
professors.exam_dates continua existindo como nota em texto livre e não é lido
aqui, justamente por não ser uma data confiável.
"""

from datetime import date
from uuid import UUID

from fastapi import APIRouter, HTTPException

from models import CalendarEventCreate, CalendarResponse
from services.auth import UserId
from services.progress import user_professor_ids
from services.supabase_client import get_supabase

router = APIRouter(prefix="/calendario", tags=["calendario"])


def _professors_by_id(user_id: str) -> dict[str, dict]:
    sb = get_supabase()
    res = sb.table("professors").select("id, name, discipline").eq("user_id", user_id).execute()
    return {row["id"]: row for row in (res.data or [])}


@router.get("", response_model=CalendarResponse)
def get_calendario(inicio: date, fim: date, user_id: UserId):
    """Atividades respondidas e eventos marcados entre `inicio` e `fim` (inclusive)."""
    professors = _professors_by_id(user_id)
    professor_ids = list(professors.keys())

    sb = get_supabase()

    activities: list[dict] = []
    if professor_ids:
        # `fim` é inclusive: o filtro vai até o fim do dia, por isso lt(fim + 1 dia).
        res = (
            sb.table("activity_results")
            .select("id, professor_id, topic, score_pct, created_at")
            .in_("professor_id", professor_ids)
            .not_.is_("score_pct", "null")
            .gte("created_at", inicio.isoformat())
            .lt("created_at", f"{fim.isoformat()}T23:59:59.999999+00:00")
            .order("created_at")
            .execute()
        )
        for row in res.data or []:
            professor = professors.get(row["professor_id"], {})
            activities.append(
                {
                    **row,
                    "professor_name": professor.get("name", "—"),
                    "discipline": professor.get("discipline", ""),
                }
            )

    events_res = (
        sb.table("calendar_events")
        .select("id, professor_id, title, kind, event_date")
        .eq("user_id", user_id)
        .gte("event_date", inicio.isoformat())
        .lte("event_date", fim.isoformat())
        .order("event_date")
        .execute()
    )
    events = [
        {
            **row,
            "professor_name": professors.get(row["professor_id"], {}).get("name"),
        }
        for row in (events_res.data or [])
    ]

    return {"activities": activities, "events": events}


@router.post("/eventos")
def create_evento(payload: CalendarEventCreate, user_id: UserId):

    if payload.professor_id is not None:
        if str(payload.professor_id) not in _professors_by_id(user_id):
            raise HTTPException(status_code=404, detail="Professor não encontrado")

    sb = get_supabase()
    res = (
        sb.table("calendar_events")
        .insert(
            {
                "user_id": user_id,
                "professor_id": str(payload.professor_id) if payload.professor_id else None,
                "title": payload.title,
                "kind": payload.kind,
                "event_date": payload.event_date.isoformat(),
            }
        )
        .execute()
    )
    if not res.data:
        raise HTTPException(status_code=500, detail="Falha ao criar o evento")
    return res.data[0]


@router.delete("/eventos/{event_id}")
def delete_evento(event_id: UUID, user_id: UserId):
    sb = get_supabase()
    res = (
        sb.table("calendar_events")
        .delete()
        .eq("id", str(event_id))
        .eq("user_id", user_id)
        .execute()
    )
    if not res.data:
        raise HTTPException(status_code=404, detail="Evento não encontrado")
    return {"deleted": str(event_id)}

File: supabase/functions/api/routes/calendario.ts
/*
  Rota de Calendário — porta de routers/calendario.py.

  GET    /calendario                  atividades respondidas + eventos numa janela
  POST   /calendario/eventos          marca um evento (prova, revisão, tarefa)
  DELETE /calendario/eventos/:id      remove um evento

  O passado vem de activity_results (o que o aluno realmente fez) e o futuro
  de calendar_events (o que ele marcou).
*/

import type { Router } from "../../_shared/router.ts";
import { currentUserId } from "../../_shared/auth.ts";
import { db } from "../../_shared/db.ts";
import { HttpError } from "../../_shared/http.ts";

const KINDS = new Set(["prova", "revisao", "tarefa", "outro"]);
const DATA_ISO = /^\d{4}-\d{2}-\d{2}$/;

function exigirData(valor: string | null, nome: string): string {
  if (!valor || !DATA_ISO.test(valor)) {
    throw new HttpError(400, `Parâmetro "${nome}" precisa ser uma data YYYY-MM-DD.`);
  }
  return valor;
}

type Professor = { id: string; name: string; discipline: string };

async function professoresPorId(userId: string): Promise<Map<string, Professor>> {
  const { data, error } = await db()
    .from("professors")
    .select("id, name, discipline")
    .eq("user_id", userId);
  if (error) throw new HttpError(500, error.message);
  return new Map((data ?? []).map((p) => [p.id as string, p as Professor]));
}

export function register(router: Router): void {
  router.get("/calendario", async (ctx) => {
    const userId = await currentUserId(ctx.req);
    const inicio = exigirData(ctx.query.get("inicio"), "inicio");
    const fim = exigirData(ctx.query.get("fim"), "fim");

    const professors = await professoresPorId(userId);
    const professorIds = [...professors.keys()];

    // deno-lint-ignore no-explicit-any
    let activities: any[] = [];
    if (professorIds.length) {
      // `fim` é inclusive: o filtro vai até o fim do dia, por isso o corte em 23:59:59.999999.
      const { data, error } = await db()
        .from("activity_results")
        .select("id, professor_id, topic, score_pct, created_at")
        .in("professor_id", professorIds)
        .not("score_pct", "is", null)
        .gte("created_at", inicio)
        .lt("created_at", `${fim}T23:59:59.999999+00:00`)
        .order("created_at");
      if (error) throw new HttpError(500, error.message);
      activities = (data ?? []).map((row) => {
        const professor = professors.get(row.professor_id as string);
        return {
          ...row,
          professor_name: professor?.name ?? "—",
          discipline: professor?.discipline ?? "",
        };
      });
    }

    const { data: eventRows, error: erroEventos } = await db()
      .from("calendar_events")
      .select("id, professor_id, title, kind, event_date")
      .eq("user_id", userId)
      .gte("event_date", inicio)
      .lte("event_date", fim)
      .order("event_date");
    if (erroEventos) throw new HttpError(500, erroEventos.message);

    const events = (eventRows ?? []).map((row) => ({
      ...row,
      professor_name: professors.get(row.professor_id as string)?.name ?? null,
    }));

    return { activities, events };
  });

  router.post("/calendario/eventos", async (ctx) => {
    const userId = await currentUserId(ctx.req);
    const payload = await ctx.body<{
      title?: unknown;
      kind?: unknown;
      event_date?: unknown;
      professor_id?: unknown;
    }>();

    if (typeof payload.title !== "string" || !payload.title.trim()) {
      throw new HttpError(400, "Campo \"title\" é obrigatório.");
    }
    const eventDate = exigirData((payload.event_date as string) ?? null, "event_date");
    const kind = (payload.kind as string | undefined) ?? "outro";
    if (!KINDS.has(kind)) {
      throw new HttpError(400, `"kind" precisa ser um de: ${[...KINDS].join(", ")}.`);
    }
    const professorId = (payload.professor_id as string | null | undefined) ?? null;

    if (professorId !== null) {
      const professors = await professoresPorId(userId);
      if (!professors.has(professorId)) throw new HttpError(404, "Professor não encontrado");
    }

    const { data, error } = await db()
      .from("calendar_events")
      .insert({
        user_id: userId,
        professor_id: professorId,
        title: payload.title,
        kind,
        event_date: eventDate,
      })
      .select();
    if (error) throw new HttpError(500, error.message);
    if (!data || !data.length) throw new HttpError(500, "Falha ao criar o evento");
    return data[0];
  });

  router.delete("/calendario/eventos/:id", async (ctx) => {
    const userId = await currentUserId(ctx.req);
    const { data, error } = await db()
      .from("calendar_events")
      .delete()
      .eq("id", ctx.params.id)
      .eq("user_id", userId)
      .select();
    if (error) throw new HttpError(500, error.message);
    if (!data || !data.length) throw new HttpError(404, "Evento não encontrado");
    return { deleted: ctx.params.id };
  });
}
